File: firefly/src/main/java/com/firefly/server/http2/router/impl/AcceptHeaderMatcher.java
package com.firefly.server.http2.router.impl;

import com.firefly.codec.http2.model.AcceptMIMEType;
import com.firefly.server.http2.router.Router;
import com.firefly.utils.CollectionUtils;
import com.firefly.utils.StringUtils;

import java.util.*;
import java.util.stream.Collectors;

import static com.firefly.codec.http2.model.MimeTypes.parseAcceptMIMETypes;

/**
 * @author Pengtao Qiu
 */
public class AcceptHeaderMatcher extends AbstractPreciseMatcher {

    @Override
    public MatchType getMatchType() {
        return MatchType.ACCEPT;
    }

    @Override
    public MatchResult match(String value) {
        if (map == null) {
            return null;
        }

        List<AcceptMIMEType> acceptMIMETypes = parseAcceptMIMETypes(value);
        if (CollectionUtils.isEmpty(acceptMIMETypes)) {
            return null;
        }
        
        for (AcceptMIMEType type : acceptMIMETypes) {
            Set<Router> set = map.entrySet().parallelStream().filter(e -> {
                String[] t = StringUtils.split(e.getKey(), '/');
                String p = t[0].trim();
                String c = t[1].trim();
                switch (type.getMatchType()) {
                    case EXACT:
                        return p.equals(type.getParentType()) && c.equals(type.getChildType());
                    case CHILD:
                        return c.equals(type.getChildType());
                    case PARENT:
                        return p.equals(type.getParentType());
                    case ALL:
                        return true;
                    default:
                        return false;
                }
            }).map(Map.Entry::getValue).flatMap(Collection::stream).collect(Collectors.toSet());
            if (!CollectionUtils.isEmpty(set)) {
                return new MatchResult(set, Collections.emptyMap(), getMatchType());
            }
        }
        return null;
    }
}

File: firefly-net/src/main/kotlin/com/fireflysource/net/http/server/impl/matcher/AcceptHeaderMatcher.kt
package com.fireflysource.net.http.server.impl.matcher

import com.fireflysource.common.string.StringUtils
import com.fireflysource.net.http.common.model.AcceptMIMEMatchType
import com.fireflysource.net.http.common.model.AcceptMIMEType
import com.fireflysource.net.http.common.model.MimeTypes.parseAcceptMIMETypes
import com.fireflysource.net.http.server.Matcher
import com.fireflysource.net.http.server.Matcher.MatchType
import com.fireflysource.net.http.server.Router
import java.util.*


class AcceptHeaderMatcher : AbstractPreciseMatcher() {

    override fun getMatchType(): MatchType? {
        return MatchType.ACCEPT
    }

    override fun match(value: String): Matcher.MatchResult? {
        if (routersMap.isEmpty()) return null

        val acceptMIMETypes = parseAcceptMIMETypes(value)
        if (acceptMIMETypes.isNullOrEmpty()) return null

        return acceptMIMETypes
            .map { findRouters(it) }
            .filter { !it.isNullOrEmpty() }
            .map { Matcher.MatchResult(it, emptyMap(), matchType) }
            .firstOrNull()
    }

    private fun findRouters(type: AcceptMIMEType) =
        routersMap.entries.filter { matchMimeType(it, type) }.flatMap { it.value }.toSortedSet()

    private fun matchMimeType(e: MutableMap.MutableEntry<String, SortedSet<Router>>, type: AcceptMIMEType): Boolean {
        val acceptType: Array<String> = StringUtils.split(e.key, '/')
        val parentType = acceptType[0].trim()
        val childType = acceptType[1].trim()
        return when (type.matchType) {
            AcceptMIMEMatchType.EXACT -> parentType == type.parentType && childType == type.childType
            AcceptMIMEMatchType.CHILD -> childType == type.childType
            AcceptMIMEMatchType.PARENT -> parentType == type.parentType
            AcceptMIMEMatchType.ALL -> true
            else -> false
        }
    }
}